pypeit/coadd3d.py

```diff
@@ -0,0 +1,1772 @@
+        if self.flux_spline is not None:
+            # Get wavelength of each pixel, and note that the WCS gives this in m, so convert to Angstroms (x 1E10)
+            numwav = vox_edges[2].size - 1
+            senswave = cube_wcs.spectral.wcs_pix2world(np.arange(numwav), 0)[0] * 1.0E10
+            sensfunc = self.flux_spline(senswave)
+
+        # Generate a datacube
+        outfile = datacube.get_output_filename("", self.cubepar['output_filename'], True, -1)
+        if self.method in ['subpixel', 'ngp']:
+            # Generate the datacube
+            wl_wvrng = None
+            if self.cubepar['save_whitelight']:
+                wl_wvrng = datacube.get_whitelight_range(np.max(self.mnmx_wv[:, :, 0]),
+                                                np.min(self.mnmx_wv[:, :, 1]),
+                                                self.cubepar['whitelight_range'])
+            if self.combine:
+                generate_cube_subpixel(outfile, cube_wcs, self.all_ra, self.all_dec, self.all_wave, self.all_sci, self.all_ivar,
+                                       np.ones(self.all_wghts.size),  # all_wghts,
+                                       self.all_spatpos, self.all_specpos, self.all_spatid, self.all_tilts, self.all_slits, self.all_align, vox_edges,
+                                       all_idx=self.all_idx, overwrite=self.overwrite, blaze_wave=self.blaze_wave,
+                                       blaze_spec=self.blaze_spec,
+                                       fluxcal=self.fluxcal, sensfunc=sensfunc, specname=self.specname, whitelight_range=wl_wvrng,
+                                       spec_subpixel=self.spec_subpixel, spat_subpixel=self.spat_subpixel)
+            else:
+                for ff in range(self.numfiles):
+                    outfile = datacube.get_output_filename("", self.cubepar['output_filename'], False, ff)
+                    ww = np.where(self.all_idx == ff)
+                    generate_cube_subpixel(outfile, cube_wcs, self.all_ra[ww], self.all_dec[ww], self.all_wave[ww], self.all_sci[ww],
+                                           self.all_ivar[ww], np.ones(self.all_wghts[ww].size),
+                                           self.all_spatpos[ww], self.all_specpos[ww], self.all_spatid[ww], self.all_tilts[ff],
+                                           self.all_slits[ff], self.all_align[ff], vox_edges,
+                                           all_idx=self.all_idx[ww], overwrite=self.overwrite, blaze_wave=self.blaze_wave,
+                                           blaze_spec=self.blaze_spec,
+                                           fluxcal=self.fluxcal, sensfunc=sensfunc, specname=self.specname,
+                                           whitelight_range=wl_wvrng,
+                                           spec_subpixel=self.spec_subpixel, spat_subpixel=self.spat_subpixel)
+
+
+def generate_image_subpixel(image_wcs, all_ra, all_dec, all_wave, all_sci, all_ivar, all_wghts,
```

Comment:
Should this go in core.datacube?

Reply:
This is easy to move to core, but it has PypeIt objects that need to be passed in. If you're OK with that, then I'm happy to move it to core, but I somehow feel that's going against the ethos of what core should represent.

Reply:
I'm fine with this being here, as it is not a class method and avoids circular imports.